gapis/trace/android/trace.go

```diff
@@ -215,32 +224,62 @@ func (t *androidTracer) Validate(ctx context.Context) error {
 
 		startFunc(ctx)
 		if !doneSignal.Wait(ctx) {
-			err = log.Err(ctx, err, "Fail to wait for done signal from Perfetto.")
+			err = log.Err(ctx, err, "Failed to wait for done signal from Perfetto.")
 		}
 		log.I(ctx, "Perfetto trace size %v bytes", written)
 	})
 	if err != nil {
-		return err
+		return nil, err
+	}
+
+	// Attempt to download perfetto trace.
+	var temp file.Path
+	if enableLocalFiles {
+		temp, err = file.TempWithExt("validation", "perfetto")
+		if err != nil {
+			return nil, log.Err(ctx, err, "Failed to create a temp file for trace")
+		}
 	}
 
 	var processor *perfetto.Processor
+	var traceLoadingErr error
+	res := &service.DeviceValidationResult{}
```

Comment:
This is not actually needed inside the closure below, so I think it would be clearer to move the construction of it below the `status.Do` call. You can then initialize the path field, when creating it:
```
res := &service.DeviceValidationResult{
  TracePath: temp.System(),
}
```